validation: More detailed on incompatible BGL

scoopr opened the pull request:
When encountering a incompatible BindGroupLayout on RenderPipelines,
attempt to report the actual difference causing the incompatibility.

**Description**
When encountering a validation error of incompatible layouts on a renderpipeline, the validation error wasn't too helpful.

```
Caused by:
    In a RenderPass
      note: encoder = `<CommandBuffer-(0, 1, Metal)>`
    In a draw command, indexed:true indirect:false
      note: render pipeline = `shadow`
    The pipeline layout, associated with the current render pipeline, contains a bind group layout at index 1 which is incompatible with the bind group layout associated with the bind group at 1
```
I shortened the final error message a little, as it was confusingly wordy, and then added info on the differences:
```
Caused by:
    In a RenderPass
      note: encoder = `<CommandBuffer-(0, 1, Metal)>`
    In a draw command, indexed:true indirect:false
      note: render pipeline = `shadow`
    The pipeline layout, associated with the current render pipeline, contains a incompatible bind group layout at index 1
      note: Entry 1 not found in assigned bindgroup layout
```

At this stage, I'm still asking opinions on how the overall approach, and implementation.

Still things to do:
- [ ] Go through the wording a bit more
- [ ] Show the labels of the BGLs
- [ ] Implement the same for compute pipelines
- [ ] .. and the checklist stuff


**Testing**
Manually munging the examples.


<!-- 
Thanks for filing! The codeowners file will automatically request reviews from the appropriate teams.

After you get a review and have addressed any comments, please explicitly re-request a review from the
person(s) who reviewed your changes. This will make sure it gets re-added to their review queue - you're no bothering us!
-->

**Checklist**

- [ ] Run `cargo fmt`.
- [ ] Run `cargo clippy`. If applicable, add:
  - [ ] `--target wasm32-unknown-unknown`
  - [ ] `--target wasm32-unknown-emscripten`
- [ ] Run `cargo xtask test` to run tests.
- [ ] Add change to `CHANGELOG.md`. See simple instructions inside file.

scoopr committed:
validation: More detailed on incompatible BGL

When encountering a incompatible BindGroupLayout on RenderPipelines,
attempt to report the actual difference causing the incompatibility.

cwfitzgerald commented on wgpu-core/src/command/bind.rs:
```diff
@@ -53,6 +53,50 @@ mod compat {
         }
+
+        fn bgl_diff(&self) -> Vec<String> {
+            let mut diff = vec![];
```
```suggestion
            let mut diff = Vec::new();
```

Just a style thing

cwfitzgerald commented on wgpu-core/src/command/bind.rs:
```diff
@@ -121,6 +165,16 @@ mod compat {
+                    return e.bgl_diff()
+                }                   
+            }
+            vec![format!("huh?!")]
```
```suggestion
            vec![String::from("huh?!")]
```

cwfitzgerald reviewed (changes requested):
Some small things, then CI, then good

scoopr committed:
Nits

scoopr committed:
fmt

scoopr committed:
Improve error messages

scoopr committed:
Implement bgl_diff for compute

scoopr commented:
Alright, I addressed the nits, tweaked the error messages, and applied the same for compute side. I think this is now in a bit better shape.

Example of with compute looks like:
```
Caused by:
    In a ComputePass
      note: encoder = `<CommandBuffer-(0, 1, Metal)>`
    In a dispatch command, indirect:false
      note: compute pipeline = `Compute pipeline`
    Incompatible bind group at index 0 in the current compute pipeline
      note: Should be compatible with bind group layout with label = `compute bgl2`
      note: Assigned bind group layout with label = `compute bgl`
      note: Entry 3 not found in assigned bindgroup layout
```

Few notes,
* I couldn't use the same labeling infra that the other errors use as I couldn't get the bgl ids nicely, but probably for the best, as I think the added context in the error messages is useful, though the lines are a bit long, and not sure if equals sign is helpful there.
* As it happens, this type of error message is presented very rarely, as lot of the other validation checks trigger before getting this far, so depending on the case, it might be nice to ease the validation on another layer to make it trigger here, but then again earlier check might have more useful context.
* And my main concern is actually that the validation check itself is different code than the validation error generation, which might cause some deviation. Its also pretty mechanical code, and it almost triggered me to go learn proc-macros to generate the diff and a fast-path validation "easily" from the same source. But that seemed a lot of effort for just this one problem :)

cwfitzgerald reviewed (approved):
Amazing!

cwfitzgerald commented:
I think we should give this a spin, then iterate if there are issues.
